pep-9999.rst

```diff
@@ -0,0 +1,217 @@
+
+    [authentication]
+
+    [authentication.pypi]
+    username = "__token__"
+    password = "pypi-AgEIcHlwaS5vcmcCJDJjMmI5NDJmLWIzNzQtNDc3ZC04ZmRkLTM4NDEzY2M1ZTU5MAACJXsicGVybWlzc2lvbnMiOiAidXNlciIsICJ2ZXJzaW9uIjogMX0AAAYgV91r8iEUUz7Q5xbsfi_KGiO3wrJR2_IOOpnFyLo62bk"
+
+    [authentication.myrepo1]
+    username = "foo"
+    password = "YmFyCg=="
+
+
+Instead of specifying the credentials in this file, Keyring can be used::
+
+    keyring set https://pypi.org/simple/ __token__
+    keyring set <repository URL> <private-repository username>
+
+
+A repository's credentials has the following attributes::
+
+    ==================  ===========  ===============================================
+    Attribute           Type         Description
+    ==================  ===========  ===============================================
+    cert                str          Path to certificate authority bundle (*.pem)
+    cert-client         str          Path to client certificate (*.pem)
+    password            str          The password for the repository, base64-encoded. When token is used, enter it here in clear-text.
+    proxy_password      str          The password for the proxy server, base64-encoded.
+    proxy_username      str          The username for the proxy server
+    username            str          The username for the repository. When token is used, set this to ``__token__``.
+    ==================  ===========  ===============================================
+
+
+Package discovery
+-----------------
+
+The PyPI repository (``repositories.pypi``) does not have to be specified in the ``config.toml``. If not specified, it will be used as the first package index by default. However, by explicitly defining it, the repository attributes can be used and it can be ordered differently.
+
+The order of the repository definitions in ``config.toml`` shall be respected when searching for packages. This allows custom package indexes to take precendence over e.g. a default PyPI repository and the use of package shadowing.
+
+
+Configuration file discovery
+----------------------------
+
+The locations of the configuration files vary slightly across platforms. You may have per-user, per-virtualenv or global (shared amongst all users) configuration:
+
+If multiple configuration files are found, then they are combined (merge and override) in the following order:
+
+1. The global file is read
+2. The per-user file is read
+3. The virtualenv-specific file is read
+
+Each file read overrides any values read from previous files.
+
+
+Global
+''''''
+
+* Unix: ``$HOME/.config/python/``
+* macOS: ``$HOME/.config/python/``
+* Windows: ``%APPDATA%\python\``
+
+
+Per-user
+''''''''
+
+* Unix: ``$HOME/.config/python/``
+* macOS: ``$HOME/.config/python/``
+* Windows: ``%APPDATA%\python\``
+
+
+Per-virtualenv
```

Comment:
I prefer "per-environment", and then define it in terms of sys.prefix. That allows for isolated non-virtual environments to contain configuration as well (e.g. a distro could include a file that defaults to their own feed rather than PyPI).

Reply:
Hm. I just realized one thing.

I think if we are picking between `$VIRTUAL_ENV` vs `sys.prefix` there's no problem going with `sys.prefix`. 

But I just realized, at least I would like to store the `config.toml` inside my git repo. Meaning, I don't want it placed in my virtual environment. Then I'd store the credentials outside of the git repo.

I wonder what the discovery could be like in this scenario, if you don't want to start making assumptions on _where_ the virtual environment is placed. For example, I usually place a `.venv` folder in the project folder. But a virtual environment could be located anywhere, really. Maybe we'd have to settle on `sys.prefix` as that problem becomes too weird to try and solve?

Reply:
I think in that situation you'll be up against the guys working on the lock file specification, as that is the one that would specify the indexes to pull from. So I doubt you'll have a project-specific configuration, because you'd have a different file with the right settings.

That said, it'll be important to align the two different files so that the lock file (or the lock file _source_ files) can use repository names, rather than full URLs, provided the name is defined in this config. And obviously you'd want to pull auth from a user-specific location.

Site configuration (aka "per environment") is really a special case - most people will never use it. But if you're embedding Python, or distributing a private copy, it becomes essential really quickly. So I'd avoid thinking of it in terms of virtualenv, because that's really not the reason for it. Save that mental energy for making things work with the other spec that's currently in motion.

Reply:
I wasn't aware of a PEP on lockfile spec was going on. Yes, definitively sounds important to sync/align there.

I updated the draft to say "per-environment" and use `sys.prefix`.